store/postgres/src/copy.rs

```diff
@@ -0,0 +1,540 @@
+                    cts::batch_size.eq(table.batch_size),
+                )
+            })
+            .collect::<Vec<_>>();
+        insert_into(cts::table).values(values).execute(conn)?;
+
+        Ok(CopyState {
+            src,
+            dst,
+            target_block,
+            tables,
+        })
+    }
+
+    fn crosses_shards(&self) -> bool {
+        self.dst.site.shard != self.src.site.shard
+    }
+
+    fn finished(&self, conn: &PgConnection) -> Result<(), StoreError> {
+        use copy_state as cs;
+
+        update(cs::table.filter(cs::dst.eq(self.dst.site.id)))
+            .set(cs::finished_at.eq(sql("now()")))
+            .execute(conn)?;
+
+        // If we imported the schema for `src`, and no other in-progress
+        // copy is using it, get rid of it again
+        if self.crosses_shards() {
+            let active_copies = cs::table
+                .filter(cs::src.eq(self.src.site.id))
+                .filter(cs::finished_at.is_null())
+                .count()
+                .get_result::<i64>(conn)?;
+            if active_copies == 0 {
+                // This is a foreign schema that nobody is using anymore,
+                // get rid of it. As a safety check (on top of the one that
+                // drop_foreign_schema does), see that we do not have
+                // metadata for `src`
+                if crate::deployment::exists(conn, &self.src.site)? {
+                    return Err(constraint_violation!(
+                        "we think we are copying {}[{}] across shards from {} to {}, but the \
+                        source subgraph is actually in this shard",
+                        self.src.site.deployment,
+                        self.src.site.id,
+                        self.src.site.shard,
+                        self.dst.site.shard
+                    ));
+                }
+                crate::catalog::drop_foreign_schema(conn, self.src.site.as_ref())?;
+            }
+        }
+        Ok(())
+    }
+}
+
+struct TableState {
+    dst_site: Arc<Site>,
+    src: Arc<Table>,
+    dst: Arc<Table>,
+    /// The `vid` of the next entity version that we will copy
+    next_vid: i64,
+    target_vid: i64,
+    batch_size: i64,
+}
+
+impl TableState {
+    fn init(
+        conn: &PgConnection,
+        dst_site: Arc<Site>,
+        src: Arc<Table>,
+        dst: Arc<Table>,
+        target_block: &EthereumBlockPointer,
+    ) -> Result<Self, StoreError> {
+        #[derive(QueryableByName)]
+        struct MaxVid {
+            #[sql_type = "diesel::sql_types::BigInt"]
+            max_vid: i64,
+        }
+
+        let target_vid = sql_query(&format!(
+            "select coalesce(max(vid), 0) as max_vid from {} where lower(block_range) <= $1",
+            src.qualified_name.as_str()
+        ))
+        .bind::<Integer, _>(&target_block.number)
+        .load::<MaxVid>(conn)?
+        .first()
+        .map(|v| v.max_vid)
+        .unwrap_or(0);
```

Comment:
If the table is empty, I believe the `target_vid` should be `-1`.

Reply:
Tru; that saves an unnecessary attempt to copy `vid = 0`